R/wrappers.R

```diff
@@ -5,7 +5,7 @@
 coalesce    = function(...) .Call(Ccoalesce, list(...), FALSE)
 setcoalesce = function(...) .Call(Ccoalesce, list(...), TRUE)
 
-fifelse = function(test, yes, no, na=NULL) .Call(CfifelseR, test, yes, no, na)
+fifelse = function(test, yes, no, na) .Call(CfifelseR, test, yes, no, if (missing(na)) substitute() else na)
 
```

Comment:
I don't agree with this change: missingness is difficult to work with. Why not just have `na=NULL` as the default then handle `NULL` as documented. Also clarifies its status as an optional argument in the `usage` section of the documentation.

Reply:
why it would be difficult to work with? could you provide an example that would impose any difficulty on user, based on current implementation

Reply:
Suppose I want to use `data.table::fifelse` in my package. Then I too have to check for missingness every time I call it. 

```r
bad_fizzbuzz <- function(number, na) {
  fifelse(number %% 15, "fizzbuzz", "", na)
}
bad_fizzbuzz(1:100) # error

bad_fizzbuzz <- function(number, na) {
  if (missing(na)) {
    fifelse(number %% 15, "fizzbuzz", "") 
  } else {
    fifelse(number %% 15, "fizzbuzz", "", na) 
  }
}


```

Reply:
I am not able to reproduce the problem. I am getting an error related to `test` argument, not `na`, which I can fix by:
```r
bad_fizzbuzz <- function(number, na) {
  fifelse(as.logical(number %% 15), "fizzbuzz", "", na)
}
bad_fizzbuzz(1:100)
```

Reply:
OK that one doesn't work. But the point is I have to be careful about `missing(na)` whenever I use a package that relies on missingness. 

Maybe this?

```r
lapply(list(list(yes = 1, no = 2, na = 3), list(yes = 2, no = 4)), function(el) fifelse(c(TRUE, FALSE, NA), el$yes, el$no, el$na))
```

Reply:
Now it is better, although it still not convincing me.
Providing 0-length object to `na` should mean that we want to skip those elements from the answer. So be equivalent to
```r
ina = which(is.na(test))
fifelse(test[!ina], yes[!ina], no[!ina])
```
Which I don't think make sense to implement due to complexity of dynamic length answer. Because of that I will try to address your use case.

Reply:
The point is not that a user really would intend to provide zero-length NA, but there are circumstances where there is no way to dynamically convey missingness in a nested function. 

Using `NULL` is clearer and more powerful. Reserve `missing()` for when an argument is truly 'missing' -- *i.e.* should have been provided but wasn't and that fact needs to be caught early -- not for defaults.

Reply:
The other advantage to using `NULL` is that it's consistent with `dplyr::if_else` and `hutils::if_else` so users can use it as a drop-in replacement.

Reply:
I amended code to support NULL. Note that setting default value to NULL would contradict manual about "must be length 1". I think it should satisfy the use case now.

Reply:
@jangorecki the manual could contain 'Must be NULL or length 1". Then it could be fully consistent with `dplyr::if_else` and `hutils::if_else`.

Reply:
I see the case for the default being`na=NA`. I think I find that clearer than `na=NULL` because when NA occurs in `test`, it is NA that is returned by default.   The manual could state that `na=NULL` is interpreted as if `na=NA`, for consistency with `dplyr::if_else` and `hutils::if_else`.

Reply:
The point of not using `na=NA` default is that it adds (very tiny) overhead. Because `is.na` have to check for each type separately. Even if there is an R C function for that it still some extra overhead comparing to `na == R_MissingArg` or `isNull(na)`. It would be handy to have type agnostic NA (julia has such NA). 
I tend to agree now that default `na=NULL` is fine as long as we document that.

Reply:
fix pushed, consistency to dplyr and hutils was already mentioned in this manual.